doxylicious/README.md

```diff
@@ -0,0 +1,229 @@
+      - adoc: Execute `true`.
+      - exec: true
+      - adoc: It is expect that this command returns without any output.
+  - step:
+      - adoc: Drink some coffee.
+verification:
+    - step: # <6>
+          - adoc: Check that coffee was drunk.
+          - exec: true
+```
+
+1. Each procedure must have a unique `id`.
+2. Published procedures¹ must have an AsciiDoc `title`.
+3. Published procedures may have an AsciiDoc `abstract`.
+4. Published procedures may have an AsciiDoc `intro`.
+5. This procedure has two steps, the first of which has three substeps. A substep:
+    * either specifies some words, using `adoc`,
+    * or specifies a command to be executed, using `exec`.
+6. Optionally, procedures may have `verification` steps.
+
+<small>¹ See the section on generating documentation for what 'published' means.</small>
+
+### A proc with prerequisites
+
+```yaml
+id: defining_prereqs 
+title:
+  adoc: Defining prerequisites
+abstract:
+  adoc: 'This procedure demonstrates prerequisites.' 
+prereqs: # <1>
+  - ref: succeeding_reliably # <2>
+    adoc: You should have already succeeded # <3>
+procedure:
+  - step: # <5>
+      - adoc: Execute `true` again
+      - exec: true
+```
+
+1. A proc provides a list of other procedures that must be done before this one. 
+2. The `ref` is the id of the prerequisite procedure.
+3. If this procedure is published, it also must define an `adoc` with the words describing the prerequisite.
+
+When `defining_prereqs` is tested the test framework will automatically execute `succeeding_reliably` first. In the terminology of software testing, each prerequisite is a [_test fixture_](https://en.wikipedia.org/wiki/Test_fixture#Software) of the procedure under test.
+
+When AsciiDoc for `defining_prereqs` is generated the prerequisites will be included.
+
+### Notional procedures
+
+Sometimes there are different ways of doing a thing. 
+For example, in describing how a piece of software might be is installed we might be able to download the binary, or we might install it using a package manager such as `dnf`.
+Subsequent procedures which describe how to use the software don't care _how_ the software was installed, only that it _was_ installed. 
+
+To model this we have the concept of a _notional_ procedure.
+
+```yaml
+id: installing_the_thing
+notional: true
+```
+
+A notional procedure doesn't define any `procedure` property and cannot itself be executed. 
+However, it can be _satisfied_ by one or more other procedures. 
+
+```yaml
+---
+id: installing_the_thing_from_the_website
+satisfies: installing_the_thing
+# ... title, procedure, etc
+---
+id: installing_the_thing_using_dnf
+satisfies: installing_the_thing
+# ... title, procedure, etc
+```
+
+Aside: `notional` and `satisfies` are conceptually the same as `interface` and `implements` in the Java programming language. Procedures already use the term `abstract`, in the sense of a summary, so different terminology is used to avoid ambiguity.
```

Comment:
continuing the parallels with java is there millage in expanding `notional` interface to include a default way of satisfying it?

I'm thinking that by default we probably want to verify that our procedures work with the proxy deployed by the operator but it would be useful to be able to inject a bare metal deployment instead. There are probably otherways of solving this.

Reply:
Who should decide what the test coverage over the satisfiers of a notional proc? Should be the person writing/controlling the notional proc? Or should it be the person writing the test? Or should it be the person executing the test? Obviously these might be the same person, but I think testing isn't going to be well served if the person writing a notional procedure gets to influence the test matrix.

TBH I haven't thought too hard about how to write procs which _aren't_ testing the operator.